src/libraries/System.Runtime/tests/System/Type/TypeTests.cs

```diff
@@ -177,21 +254,874 @@ public void GetTypeFromDefaultHandle()
+                new object[] { typeof(UInt64Enum) },
+
+                // Array, pointers
+                new object[] { typeof(string[]) },
+                new object[] { typeof(int[]) },
+                new object[] { typeof(int[][]) },
+                new object[] { typeof(int).MakeArrayType(1) },
+                new object[] { typeof(int[,]) },
+                new object[] { typeof(int*) },
+                new object[] { typeof(int).MakeByRefType() },
+
+                // Classes, structs, interfaces, enums
+                new object[] { typeof(NonGenericClass) },
+                new object[] { typeof(NonGenericStruct) },
+                new object[] { typeof(NonGenericInterface) },
+                new object[] { typeof(AbstractClass) },
+                new object[] { typeof(StaticClass) },
+
+                // Generic types.
+                new object[] { typeof(GenericClass<>).GetTypeInfo().GetGenericArguments()[0] },
+
+                // By-Ref Like.
+                new object[] { typeof(TypedReference) },
+                new object[] { typeof(ArgIterator) },
+                new object[] { typeof(RuntimeArgumentHandle) },
+            };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetGenericTypeDefinition_NotGenericType_TestData))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/36642", TestRuntimes.Mono)]
+        public void GetGenericTypeDefinition_InvokeNotGenericType_ThrowsInvalidOperationException(Type t)
+        {
+            Assert.Throws<InvalidOperationException>(() => t.GetGenericTypeDefinition());
+        }
+
+        public static IEnumerable<object[]> HasElementType_TestData()
+        {
+            return new object[][]
+            {
+                // Primitives
+                new object[] { typeof(string), false },
+                new object[] { typeof(sbyte), false },
+                new object[] { typeof(byte), false },
+                new object[] { typeof(short), false },
+                new object[] { typeof(ushort), false },
+                new object[] { typeof(int), false },
+                new object[] { typeof(uint), false },
+                new object[] { typeof(long), false },
+                new object[] { typeof(ulong), false },
+                new object[] { typeof(char), false },
+                new object[] { typeof(bool), false },
+                new object[] { typeof(float), false },
+                new object[] { typeof(double), false },
+                new object[] { typeof(IntPtr), false },
+                new object[] { typeof(UIntPtr), false },
+                new object[] { typeof(void), false },
+                new object[] { typeof(object), false },
+                new object[] { typeof(Array), false },
+                new object[] { typeof(ValueType), false },
+                new object[] { typeof(Delegate), false },
+                new object[] { typeof(Enum), false },
```

Comment:
It looks like there's a lot of duplication being added as part of these member datas, e.g. each one has a Primitives, a Primitives enums, etc.  Tests can have multiple [MemberData(...)] attributes, in which case they'll use the union of all of them.  Can these common sections be extracted into their own methods to avoid the duplication?

Reply:
So unfortunately this is by-design because every pretty much every single one of these types that we're providing as test data is different from each other type by at least one property. I suppose I could experiment with something like
Enums_MemberData` or `Primitives_MemberData`? But I don't know how I could do this in a very readable way.

I did experiment a while back with a sort of test harness (similar to that for ICollection)

E.g.
```
private class TypeTestBase
{
     protected abstract bool IsValueType { get; }

     protected virtual bool IsPointer { get; } = false;

    [Fact]
    public void IsPointer_Get_returnsExpected()
    {
        Assert.Equal(IsPointer, Type.IsPointer);
    }

    [Fact]
    public void IsValueType_Get_returnsExpected()
    {
        Assert.Equal(IsValueType, Type.IsValueType);
    }
}

public class Int32TypeTests : TypeTestBase
{
     public override Type Type => typeof(int);

     protected override bool IsPointer => true;
}
```

Reply:
Thanks, @hughbe.  I'm not quite following.  Why can't you do something like:
```C#
public static IEnumerable<object[]> Primitives_False_TestData() =>
    from primitive in Primitives_TestData()
    select new object[] { primitive[0], false };

public static IEnumerable<object[]> Primitives_Null_TestData() =>
    from primitive in Primitives_TestData()
    select new object[] { primitive[0], null };

public static IEnumerable<object[]> Primitives_TestData()
{
    return new object[][]
    {
                new object[] { typeof(string) },
                new object[] { typeof(sbyte) },
                new object[] { typeof(byte) },
                new object[] { typeof(short) },
                new object[] { typeof(ushort) },
                new object[] { typeof(int) },
                new object[] { typeof(uint) },
                new object[] { typeof(long) },
                new object[] { typeof(ulong) },
                new object[] { typeof(char) },
                new object[] { typeof(bool) },
                new object[] { typeof(float) },
                new object[] { typeof(double) },
                new object[] { typeof(IntPtr) },
                new object[] { typeof(UIntPtr) },
                new object[] { typeof(void) },
                new object[] { typeof(object) },
                new object[] { typeof(Array) },
                new object[] { typeof(ValueType) },
                new object[] { typeof(Delegate) },
                new object[] { typeof(Enum) },
    }
};
```
and then the same for the other groupings, and then on the tests just have multiple member datas, e.g.
```C#
[Theory]
[MemberData(nameof(Primitives_False_TestData))]
[MemberData(nameof(PrimitivesEnums_False_TestData))]
[MemberData(nameof(ArraysPointers_False_TestData))]
...
public void ContainsGenericParameters_Get_ReturnsExpected(Type t, bool expected)
{
    ...
}
```
Is that problematic?  I'm not sure what you mean by "every single one of these types that we're providing as test data is different from each other type by at least one property".

Reply:
I think actually that these tests may all be duplicated anyways, looking at `TypePropertyTests`. So we may not need to do this

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime/tests/System/Type/TypePropertyTests.cs

Will investigate tomorrow

Reply:
Yeah a lot of this was already duplicated in TypePropertyTests. Was able to add a few virtual members to `TypePropertyTestBase` and also add some more coverage for delegates, by-ref types etc. and avoid the need for all this `MemberData` stuff!